### ISSUE TITLE
Use unique ip address per agent pool

### ISSUE THREAD

**Issue by PascalVA** [Azure/aks-engine#3450]
At the moment when deploying an aks-engine cluster with VMSS for the agent pools, only one outbound ip address is created for all of the agent pools to share. It would be nice if we could at least have the option to have one outbound ip address per agent pool. (This would be especially useful for multi-tenant clusters.

**Explain why AKS Engine needs it**
When hosting multiple customers on a single cluster it is required for security to have a separate outbound ip address. If the pools need access to firewall-ed resources (DBs, resources, CDNs, etc...), you otherwise have no options to limit access for the other pool nodes.

**Describe the solution you'd like**
aks-engine should implement a flag like **AgentPoolOutboundIPs**.

This would require the following settings:
OutboundIps: 0
DisableOutboundSNAT: true

Basically, if you would have three agent pools named **pool01**, **pool02** and **pool03**. It should make the following configurations:
https://pastebin.com/r3a3WbWe

I hope it is a clear, it is a bit hard to explain in an issue.

**Describe alternatives you've considered**
The only alternative we have considered (and implemented) is to do it manually after aks-engine generate, which is not so nice. Especially when we then try to scale, it conflicts with the frontend ips and loadbalancing rules created by kubernetes loadbalancers.

**Comment by jackfrancis** [comment]
Hi @PascalVA, will this help?:

https://github.com/Azure/aks-engine/pull/3085

**Comment by jackfrancis** [comment]
https://github.com/Azure/aks-engine/pull/3085 has already merged btw, the `loadBalancerOutboundIPs` configuration has been available starting w/ AKS Engine v0.50.0. If this works for you, test w/ the latest release and report back!

**Comment by PascalVA** [comment]
Hi @jackfrancis, Thanks for the quick reply!

I tried this before (and am now actually setting it to 0). Unless I misunderstood something, this option does not actually solve the problem we are facing. When I set the **loadBalancerOutboundIPs** to **3** for example, the ip's were not distributed by nodepool, but rather randomly per node (or maybe even changing periodically, I don't remember). What we really need is for each node pool to have a fixed ip for all of it's instances. (So one backendAddressPool per vmss with one outbound rule that references one ip address for that node). It would also be nice if the ips could be named something like: POOL_NAME-outbound for clarity.

If I missed something on setting this up, please point me to the proper documentation :smile:

**Comment by jackfrancis** [comment]
@aramase @khenidak @feiskyer can either of you confirm whether it is possible to run Kubernetes on Azure w/ per-node outbound IP address SNAT'ing? I'm not aware of that ever having been implemented.

@PascalVA in the meantime, can you identify the particular traffic with HTTP headers (or the like, depending on the protocol)?

**Comment by PascalVA** [comment]
@jackfrancis We are currently configuring the outbound rules, backendaddress pools and frontend IP configurations ourselves. It's a bit of a pain, but it does the job. I thought this would be an important enough use case to implement in aks-engine however.

If you have any suggestions about how to handle our deployments failing when there are frontend ipconfigurations from the k8s loadbalancers, that would be great though!

**Comment by dmeytin** [comment]
+1 for this ask to prevent SNAT port exhaustion in a big clusters

**Comment by PascalVA** [comment]
> @jackfrancis We are currently configuring the outbound rules, backendaddress pools and frontend IP configurations ourselves. It's a bit of a pain, but it does the job. I thought this would be an important enough use case to implement in aks-engine however.
> 
> If you have any suggestions about how to handle our deployments failing when there are frontend ipconfigurations from the k8s loadbalancers, that would be great though!

Just to follow up on this. We now first query the LBs with ansible and build the list of frontend ip configurations ourselves.
I don't think I have time myself anytime soon to dig into the source of aks-engine and make a PR myself though.

**Comment by gabrielmcf** [comment]
This is also an use case we would like to see! We're experiencing lots os problems with SNAT port exhaustion and have some applications that needs an fixed outbound IP, so we thought about using one per node pool. I wasn't able to figure out how to configure this by hand on the portal.

**Comment by bridgetkromhout** [comment]
Hi, @gabrielmcf @PascalVA @dmeytin - writing back to you on this issue because AKS Engine is deprecated. Are you able to solve for your use cases with AKS or with CAPZ? Please take a look at https://github.com/Azure/aks-engine/#project-status and let us know if we can answer any questions; thanks.

**Comment by PascalVA** [comment]
Hi @bridgetkromhout!

Thanks for the update. We ended up just overriding the settings in the arm templates generated by AKS-engine and have been running with unique outbound ip addresses for almost two years now. 

We are going to look into CAPZ for future deployments and see if we can replicate what we have.

I will close the issue since it will never be addressed.

Thanks!
